### PR TITLE
Update GDA Implementation

### PR DESCRIPTION
**Pull request by cygaar**
Updated the logic + added a set of unit tests.

Some major changes:
1) The new spot price will now be divided by the time decay factor to prevent overflows for buyInfo
2) A similar setup will be used for sellInfo - we will divide by alpha^n for the new spotPrice
3) Added a full set of unit tests
4) Switched to using PRBMath for all GDA operations to be consistent

### CODE REVIEW

File: src/bonding-curves/GDACurve.sol [hunk 1]
```diff
@@ -56,90 +49,65 @@ contract GDACurve is ICurve, CurveErrorCodes {
...
-        (uint256 alpha, , ) = _parseDelta(delta);
+        uint256 spotPrice_ = uint256(spotPrice);
+        (uint256 alpha,,) = _parseDelta(delta);
+        uint256 alphaPowN = uint256(alpha).powu(numItems);
 
-        uint256 alphaPowN = uint256(alpha).fpow(
-            numItems,
-            FixedPointMathLib.WAD
-        );
+        // TODO: should we cap the time elapsed or decay factor?
```
Comment (cygaar) [inline src/bonding-curves/GDACurve.sol:64]: cc @sudo-owen

File: src/bonding-curves/GDACurve.sol [hunk 2]
```diff
@@ -151,95 +119,85 @@ contract GDACurve is ICurve, CurveErrorCodes {
...
-        // and spotPrice <= type(uint128).max
-        newSpotPrice = uint128(
-            uint256(spotPrice).fmul(invAlphaPowN, FixedPointMathLib.WAD)
-        );
-        if (newSpotPrice < MIN_PRICE) {
-            newSpotPrice = uint128(MIN_PRICE);
-        }
+        uint256 spotPrice_ = uint256(spotPrice);
+        (uint256 alpha,,) = _parseDelta(delta);
+        // TODO: this value may overflow, should we cap the value?
```
Comment (cygaar) [inline src/bonding-curves/GDACurve.sol:131]: cc @sudo-owen

File: src/bonding-curves/GDACurve.sol (hunk 1 above)
Comment (sudo-owen) [reply src/bonding-curves/GDACurve.sol:62]: We could cap the max decay factor to 1/100 perhaps? 

Would be enough to capture the range for most use cases and avoids issues with the price getting too low.

File: src/bonding-curves/GDACurve.sol (hunk 2 above)
Comment (cygaar) [reply src/bonding-curves/GDACurve.sol]: True, but we will still need a cap on the time for this direction otherwise the price will be too large

Comment (sudo-owen) [reply src/bonding-curves/GDACurve.sol]: We want to limit the `boostFactor` right, not necessarily time elapsed? e.g. if lambda is very very small, e.g. 0.5% a day, then even after 100 days, the price could still be valid

Comment (cygaar) [reply src/bonding-curves/GDACurve.sol]: Yeah, that's what I meant

Comment (sudo-owen) [reply src/bonding-curves/GDACurve.sol]: Makes sense. Thoughts on 1/100 and 100x being sane ceilings for the buy and sell side?

Comment (cygaar) [reply src/bonding-curves/GDACurve.sol]: It's hard to tell - I currently have it set to 1024 in my latest PR, but I still need to think about the potential effects of other values

**Pull request merged**